### PR TITLE
[scheduler] Improve the performance of the recurring-events utils

### PR DESCRIPTION
**Pull request by flaviendelangle**

### CODE REVIEW

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts [hunk 1]
```diff
@@ -1,221 +1,387 @@
...
-  adapter: Adapter,
-): (date: TemporalSupportedObject) => boolean {
-  const hasCount = typeof rule.count === 'number' && rule.count > 0;
-  const hasUntil = !!rule.until;
-
-  if (hasCount && hasUntil) {
-    throw new Error(
-      'Scheduler: The recurring rule cannot have both the count and until properties.',
-    );
-  }
```
Comment (rita-codes) [inline packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:90]: question: Are we checking this in another place? (maybe now it's not needed)

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts [hunk 2]
```diff
@@ -1,221 +1,387 @@
...
-      // Only exact "same month + same day" recurrence is supported.
-      // Any use of BYMONTH, BYMONTHDAY, BYDAY, or multiple values is not allowed.
-      if (rule.byMonth?.length || rule.byMonthDay?.length || rule.byDay?.length) {
-        throw new Error(
-          'Scheduler: The yearly recurrences only support exact same date recurrence (month/day of DTSTART).',
-        );
+      // The recurrence event has ended.
+      if (this.rule.count !== undefined && remainingCount <= 0) {
+        break;
       }
```
Comment (rita-codes) [inline packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:375]: question: Are we throwing now if the yearly has one of these rules? Just wanted to avoid silent errors 👀

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts [hunk 3]
```diff
@@ -190,351 +186,306 @@ describe('recurring-events/getRecurringEventOccurrencesForVisibleDays', () => {
     });
   });
```
Comment (rita-codes) [inline packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts:188]: As a side note: before the refactor there were a few smaller helpers that were really easy to test in isolation, which also made edge-cases and regressions easier to spot. With the new structure, some of those scenarios now require a bit more setup to validate. Not a big deal, just something I noticed while reviewing the tests 🧐

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts [hunk 4]
```diff
@@ -190,351 +186,306 @@ describe('recurring-events/getRecurringEventOccurrencesForVisibleDays', () => {
...
-              rule,
-              adapter.addDays(adapter.addMonths(start, 1), 1),
-              adapter,
-              event,
-            ),
-          ).to.equal(false);
-        });
-      });
-
-      describe('byDay ordinals', () => {
```
Comment (rita-codes) [inline packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts:353]: I see a lot of tests deleted here 👀 I assume we're covering all of these cases, right? or some of them are not necessary anymore

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts [hunk 5]
```diff
@@ -1,221 +1,387 @@
...
+    this.interval = Math.max(1, this.rule.interval ?? 1);
+
+    // Adjust scan range to catch multi-day events starting before visible range
+    const eventDuration = getEventDurationInDays(adapter, event);
+    this.scanFirstDay = adapter.startOfDay(adapter.addDays(start, 1 - eventDuration));
+    this.scanLastDay = adapter.startOfDay(end);
+
+    // Pre-compute boundaries and exclusions
+    this.exDateKeys = new Set(event.exDates?.map((d) => getDateKey(d, adapter)));
+    this.untilBoundary = this.rule.until ? adapter.startOfDay(this.rule.until) : null;
```
Comment (rita-codes) [inline packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:85]: The until day is inclusive, shouldn't be in this case `adapter.endOfDay(this.rule.until)`?

**Review by rita-codes — APPROVED** [review]
Left a couple of comments and questions, it looks pretty good! SO much smarter 🥇 thanks for doing this!! 👏

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts (hunk 1 above)
Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:90]: You are right, the current behavior just stops on the stricter of the two criteria, but if the spec this is not valid, I shoudl through.
I'll add it back :+1:

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts (hunk 2 above)
Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:375]: Could you elaborate please?

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts (hunk 4 above)
Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts:353]: I migrated some of them to use `getRecurringEventOccurrencesForVisibleDay` directly and some of them we not testing anything more than existing ones.

But if you find a test that I deleted and is useful, I can migrate it :+1:

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts (hunk 5 above)
Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:85]:
```ts
      // The occurrence is beyond the UNTIL boundary.
      if (this.untilBoundary !== null && this.adapter.isAfter(currentDay, this.untilBoundary)) {
        break;
      }
```

We do all of our comparisons with days whose times are `00:00:00` so `adapter.isAfter` does not fire when both dates are exactly equal.
I think the current approach is safe :+1:

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts (hunk 3 above)
Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts:188]: Yes it's the trade off here.
It removed `getEndGuard` and `matchesOccurrence`.

If you think it's worth it, I can try to test the method that jump from one occurrence to the next manually.
But I think it's useful to test the whole pipeline a lot as well here, to avoid having each piece work correctly, but the whole thing broken because we connect them correctly.

Comment (rita-codes) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.test.ts:188]: Yes, agree, we just need to pay more attention to these tests so we are sure we're testing all the scenarios 🙌

File: packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts (hunk 2 above)
Comment (rita-codes) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:375]: Yes, I was talking about this prev code:
```
// Only exact "same month + same day" recurrence is supported.
      // Any use of BYMONTH, BYMONTHDAY, BYDAY, or multiple values is not allowed.
      if (rule.byMonth?.length || rule.byMonthDay?.length || rule.byDay?.length) {
        throw new Error(
          'Scheduler: The yearly recurrences only support exact same date recurrence (month/day of DTSTART).',
        );
```
We should throw if the yearly rrule has byMonth/byDay or byMonthDay, no? I couldn’t find where we're doing that 👀

Comment (flaviendelangle) [reply packages/x-scheduler-headless/src/utils/recurring-events/getRecurringEventOccurrencesForVisibleDays.ts:375]: Indeed, I added it back the throw 👍
